example/search/search_replicated2.c

```diff
@@ -146,13 +148,16 @@ create_p4est (search_partition_global_t *g)
     p4est_new_ext (sc_MPI_COMM_WORLD, g->conn, 0, g->uniform_level, 1, 0,
                    NULL, g);
 
-  /* to do: better make a loop of non-recursive refinement and partition.
-     even when there is no refinement at all, execute partition at least once:
-     this is necessary since p4est_new_ext does not partition for coarsening */
+  /* We like the invariant that after partitioning, partition-independent
+     coarsening is always possible since every family of siblings is placed
+     on a single process; this is not guaranteed by p4est_new_ext. */
+  p4est_partition (g->p4est, 1, NULL);
 
   /* refine the forest adaptively around two points g->a and g->b */
-  p4est_refine (g->p4est, 1, refine_fn, NULL);
-  p4est_partition (g->p4est, 0, NULL);
+  for (il = g->uniform_level; il < g->max_level; il++) {
+    p4est_refine (g->p4est, 0, refine_fn, NULL);
```

Comment:
You may break the loop early if the global number of quadrants does not change on refinement. This may happen if for example the refine callback finds no reason to refine further at some intermediate level.